R/MetaAnalysis.R

```diff
@@ -0,0 +1,424 @@
+#' @param method          The meta-analysis method to use.  Options are "bayesian" or "DL" (DerSimonian-Laird).
+#' @param addTraditional  Boolean indicating if traditional meta-analysis (i.e., "DL") results should be added to result (if method is "bayesian").
+#'
+#' @return
+#' Does not return a value, but creates a new zip file in the \code{maExportFolder} for the meta-analyses.
+#'
+#' @export
+synthesizeResults <- function(allDbsFolder,
+                              maExportFolder = allDbsFolder,
+                              maxCores,
+                              method = "bayesian",
+                              resultsZipPattern = "^Results_.*\\.zip",
+                              addTraditional = TRUE) {
+  
+  if (!file.exists(maExportFolder)) {
+    dir.create(maExportFolder, recursive = TRUE)
+  }
+  
+  ParallelLogger::addDefaultFileLogger(file.path(maExportFolder, "metaAnalysisLog.txt"))
+  on.exit(ParallelLogger::unregisterLogger("DEFAULT_FILE_LOGGER", silent = TRUE))
+  
+  if (!file.exists(allDbsFolder)) {
+    stop(sprintf("The allDbsFolder path does not exist:\n\t", allDbsFolder))
+  }
+  
+  if (!(method %in% c("bayesian", "DL"))) {
+    stop("Accepted method arguments are \"bayesian\" or \"DL\"")
+  }
+  
+  resultsSets <- list.files(path = allDbsFolder, pattern = resultsZipPattern)
+  ParallelLogger::logInfo(sprintf("Found %d zip files matching pattern %s for synthesizing", length(resultsSets), resultsZipPattern))
+  if(length(resultsSets) == 0) {
+    stop(sprintf("No results found matching pattern %s in directory %s", resultsZipPattern, allDbsFolder))
+  } else if(length(resultsSets) == 1) {
+    stop(sprintf("Only single result set found matching pattern %s in directory %s", resultsZipPattern, allDbsFolder))
+  }
+  
+  mainResults <- lapply(resultsSets, loadDatabaseResults, allDbsFolder = allDbsFolder)
+  mainResults <- do.call(rbind, mainResults)
+  mainResults <- split(mainResults, paste(mainResults$targetId, mainResults$comparatorId, mainResults$analysisId))
+  
+  if(method == "bayesian") {
+    profiles <- lapply(resultsSets, loadLikelihoodProfiles, allDbsFolder = allDbsFolder)
+    profiles <- do.call(rbind, profiles)
+    profiles <- split(profiles, paste(profiles$targetId, profiles$comparatorId, profiles$analysisId))
+    
+    groups <- lapply(names(mainResults), function(name) list(mainResults = mainResults[[name]], profiles = profiles[[name]]))
+    
+    rm(mainResults)
+    rm(profiles)
+    
+  } else {
+    groups <- lapply(names(mainResults), function(name) list(mainResults = mainResults[[name]]))
+  }
+  
+  ParallelLogger::logInfo("Performing cross-database evidence synthesis")
+  cluster <- ParallelLogger::makeCluster(min(maxCores, 10))
+  indices <- 1:length(groups)
+  results <- ParallelLogger::clusterApply(cluster, indices, SkeletonComparativeEffectStudy:::computeGroupMetaAnalysis, groups, method, addTraditional)
```

Comment:
Is there a particular reason why you're iterating over the indices, and sending all the groups to all nodes, instead of just iterating over the groups? The latter would require less communication with the nodes, so
```r
results <- ParallelLogger::clusterApply(cluster, groups, SkeletonComparativeEffectStudy:::computeGroupMetaAnalysis, method, addTraditional)
```

Reply:
Yes, should have changed this back to sending each respective group to the nodes. I had changed it to the indices to troubleshoot an issue I had at one point and needed to pin down which group was the culprit (due to NAs being returned by EvidenceSynthesis).  But changed it back to applying over groups.